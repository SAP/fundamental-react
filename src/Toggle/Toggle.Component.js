import React from 'react';
import { Toggle } from '../';
import { Description, DocsText, DocsTile, Header, Import, Properties, Separator } from '../_playground';

export const ToggleComponent = () => {
    const toggleCode = `<Toggle size="xs" id="Yj07w604">
    Extra Small toggle
</Toggle>
<Toggle size="s" id="Yj07w605" checked>
    Small toggle
</Toggle>
<Toggle id="Yj07w606">Normal toggle</Toggle>
<Toggle size="l" id="Yj07w607" checked>
    Large toggle
</Toggle>`;

    const toggleDisabledCode = `<Toggle size="xs" id="Yj07w608" disabled>
    Extra Small toggle
</Toggle>
<Toggle size="s" id="Yj07w609" checked disabled>
    Small toggle
</Toggle>
<Toggle id="Yj07w610" disabled>
    Normal toggle
</Toggle>
<Toggle size="l" id="Yj07w611" checked disabled>
    Large toggle
</Toggle>`;

    return (
        <div>
            <Header>Toggle</Header>
            <Description>
                The toggle component is used to activate or deactivate an element. Uses a visual metaphor that is know
                to the user with visible differences between on and off state. It is recommended to always display the
                toggle with a label above it as well as the label of the selected state. For example, the label above
                would be Active, the toggle state would be “on” and the selected state label displayed to the right of
                the toggle would be “Yes”.
            </Description>
            <Import module='Toggle' path='/fundamental-react/src/' />

            <Separator />

            <Properties
                properties={[
                    { name: 'size', description: 'string - The size of the toggle.' },
                    { name: 'id', description: 'string - The id of the toggle.' }
                ]}
                type='Inputs' />

            <Separator />

            <h2>Toggle Sizes</h2>
            <Description>
                The toggle can be set to 4 sizes: 'xs', 's', 'normal' and 'l'. For 'normal' size leave empty. <br />
                When used with forms, it is recommended to use the small size so that form components will be
                consistent.
            </Description>
            <DocsTile>
                <Toggle id='Yj07w604' size='xs'>
                    Extra Small toggle
                </Toggle>
                <Toggle checked id='Yj07w605'
                    size='s'>
                    Small toggle
                </Toggle>
                <Toggle id='Yj07w606'>Normal toggle</Toggle>
                <Toggle checked id='Yj07w607'
                    size='l'>
                    Large toggle
                </Toggle>
            </DocsTile>
            <DocsText>{toggleCode}</DocsText>

            <Separator />

            <h2>Disabled state</h2>
            <Description>
                The toggle can be set to the disable state and still indicate the state of the toggle. To do this, add
                the 'disabled' parameter to the element.
            </Description>
            <DocsTile>
                <Toggle disabled id='Yj07w608'
                    size='xs'>
                    Extra Small toggle
                </Toggle>
                <Toggle checked disabled
                    id='Yj07w609' size='s'>
                    Small toggle
                </Toggle>
                <Toggle disabled id='Yj07w610'>
                    Normal toggle
                </Toggle>
                <Toggle checked disabled
                    id='Yj07w611' size='l'>
                    Large toggle
                </Toggle>
            </DocsTile>
            <DocsText>{toggleDisabledCode}</DocsText>

            <Separator />
        </div>
    );
};
