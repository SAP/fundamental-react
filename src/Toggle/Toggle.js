import React from 'react';
import PropTypes from 'prop-types';

export class Toggle extends React.Component {
    constructor(props) {
        super(props);
        this.state = { checked: props.checked };
        this.handleChange = this.handleChange.bind(this);
    }

    handleChange() {
        this.setState({ checked: !this.state.checked });
    }

    render() {
        const { size, id, disabled, children, className, ...rest } = this.props;

        return (
            <div className={`fd-form__item fd-form__item--check${className ? ' ' + className : ''}`} {...rest}>
                <label className='fd-form__label' htmlFor={id}>
                    <span className={`fd-toggle${size ? ' fd-toggle--' + size : ''} fd-form__control`}>
                        <input
                            type='checkbox'
                            id={id}
                            checked={this.state.checked}
                            onChange={this.handleChange}
                            disabled={disabled} />
                        <span className='fd-toggle__switch' role='presentation' />
                    </span>
                    {children}
                </label>
            </div>
        );
    }
}

Toggle.propTypes = {
    className: PropTypes.string,
    disabled: PropTypes.bool,
    id: PropTypes.string,
    size: PropTypes.oneOf(['', 'xs', 's', 'l'])
};
