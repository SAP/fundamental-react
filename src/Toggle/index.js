export { default as Toggle } from './Toggle.js';
