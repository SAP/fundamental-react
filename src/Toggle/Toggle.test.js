import React from 'react';
import renderer from 'react-test-renderer';
import { Toggle } from './Toggle';
import { mount, shallow } from 'enzyme';

describe('<Toggle />', () => {
    const defaultToggle = <Toggle>Normal toggle</Toggle>;
    const checkedToggle = (
        <Toggle checked className='blue'
            size='s'>
            Small toggle
        </Toggle>
    );
    const disabledToggle = <Toggle disabled>Normal toggle</Toggle>;
    const disabledCheckedToggle = (
        <Toggle checked disabled>
            Normal toggle
        </Toggle>
    );

    test('create toggle component', () => {
        // Default toggle
        let component = renderer.create(defaultToggle);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // checked toggle
        component = renderer.create(checkedToggle);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // disabled toggle
        component = renderer.create(disabledToggle);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // checked disabled toggle
        component = renderer.create(disabledCheckedToggle);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('Toggle state change', () => {
        const wrapper = shallow(defaultToggle);

        // check that toggle is not checked
        expect(wrapper.state(['checked'])).toBeFalsy();

        // simulate changing toggle
        wrapper.find('input[type="checkbox"]').simulate('change');

        // check that toggle is checked
        expect(wrapper.state(['checked'])).toBeTruthy();
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the Toggle component', () => {
            const element = mount(<Toggle data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        xtest('should allow props to be spread to the Toggle component\'s label element', () => {
            // TODO: placeholder for this test description once that functionality is built
        });

        xtest('should allow props to be spread to the Toggle component\'s input element', () => {
            // TODO: placeholder for this test description once that functionality is built
        });
    });
});
