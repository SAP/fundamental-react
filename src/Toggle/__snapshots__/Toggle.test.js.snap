// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> create toggle component 1`] = `
<div
  className="fd-form__item fd-form__item--check"
>
  <label
    className="fd-form__label"
  >
    <span
      className="fd-toggle fd-form__control"
    >
      <input
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="fd-toggle__switch"
        role="presentation"
      />
    </span>
    Normal toggle
  </label>
</div>
`;

exports[`<Toggle /> create toggle component 2`] = `
<div
  checked={true}
  className="fd-form__item fd-form__item--check blue"
>
  <label
    className="fd-form__label"
  >
    <span
      className="fd-toggle fd-toggle--s fd-form__control"
    >
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="fd-toggle__switch"
        role="presentation"
      />
    </span>
    Small toggle
  </label>
</div>
`;

exports[`<Toggle /> create toggle component 3`] = `
<div
  className="fd-form__item fd-form__item--check"
>
  <label
    className="fd-form__label"
  >
    <span
      className="fd-toggle fd-form__control"
    >
      <input
        disabled={true}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="fd-toggle__switch"
        role="presentation"
      />
    </span>
    Normal toggle
  </label>
</div>
`;

exports[`<Toggle /> create toggle component 4`] = `
<div
  checked={true}
  className="fd-form__item fd-form__item--check"
>
  <label
    className="fd-form__label"
  >
    <span
      className="fd-toggle fd-form__control"
    >
      <input
        checked={true}
        disabled={true}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="fd-toggle__switch"
        role="presentation"
      />
    </span>
    Normal toggle
  </label>
</div>
`;
